stats/bridge_stats.go

```diff
@@ -0,0 +1,347 @@
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	// collect bridge stats and send to ronin stats
+	info := &BridgeInfo{
+		Node:           s.node,
+		Operator:       s.operator,
+		Version:        s.version,
+		LastError:      s.lastError,
+		ProcessedBlock: s.processedBlock,
+	}
+
+	// count pending/failed tasks from database
+	pending, err := s.store.CountTasks(s.chainId, "pending")
+	if err != nil {
+		log.Error("error while getting pending tasks", "err", err)
+	} else {
+		info.PendingTasks = int(pending)
+	}
+
+	failed, err := s.store.CountTasks(s.chainId, "failed")
+	if err != nil {
+		log.Error("error while getting failed tasks", "err", err)
+	} else {
+		info.FailedTasks = int(failed)
+	}
+
+	report := map[string][]interface{}{
+		"emit": {"bridge-stats", info},
+	}
+	return conn.WriteJSON(report)
+}
+
+func (s *Service) loop(conn *connWrapper) {
+	sendStatsTicker := time.NewTicker(10 * time.Second)
+	for {
+		select {
+		case msg := <-s.errCh:
+			s.setLastError(msg.Listener, msg.Err)
+		case msg := <-s.processedBlockCh:
+			s.setProcessedBlock(msg.Listener, msg.ProcessedBlock)
+		case <-sendStatsTicker.C:
+			if err := s.report(conn); err != nil {
+				log.Warn("bridge stats report failed", "err", err)
+			}
+		}
+	}
+}
+
+// readLoop loops as long as the connection is alive and retrieves data packets
+// from the network socket. If any of them match an active request, it forwards
+// it, if they themselves are requests it initiates a reply, and lastly it drops
+// unknown packets.
+func (s *Service) readLoop(conn *connWrapper) {
+	// If the read loop exits, close the connection
+	defer conn.Close()
+
+	for {
+		// Retrieve the next generic network packet and bail out on error
+		var blob json.RawMessage
+		if err := conn.ReadJSON(&blob); err != nil {
+			log.Warn("Failed to retrieve stats server message", "err", err)
+			return
+		}
+		// If the network packet is a system ping, respond to it directly
+		var ping string
+		if err := json.Unmarshal(blob, &ping); err == nil && strings.HasPrefix(ping, "primus::ping::") {
+			if err := conn.WriteJSON(strings.Replace(ping, "ping", "pong", -1)); err != nil {
+				log.Warn("Failed to respond to system ping message", "err", err)
+				return
+			}
+			continue
+		}
+		// Not a system ping, try to decode an actual state message
+		var msg map[string][]interface{}
+		if err := json.Unmarshal(blob, &msg); err != nil {
+			log.Warn("Failed to decode stats server message", "err", err)
+			return
+		}
+		log.Trace("Received message from stats server", "msg", msg)
+		if len(msg["emit"]) == 0 {
+			log.Warn("Stats server sent non-broadcast", "msg", msg)
+			return
+		}
+		command, ok := msg["emit"][0].(string)
+		if !ok {
+			log.Warn("Invalid stats server message type", "type", msg["emit"][0])
+			return
+		}
+		// If the message is a ping reply, deliver (someone must be listening!)
+		if len(msg["emit"]) == 2 && command == "node-pong" {
+			select {
+			case s.pongCh <- struct{}{}:
```

Comment:
There is no consumer on this channel, so the second push to this channel fails, falls to default case and this function returns. Is this an expected behavior?

Reply:
hmm..actually this is unused, I will remove it now